src/lib/common/storage.ts

```diff
@@ -0,0 +1,77 @@
+const storageQuotaErrorMessage =
+  "Unable to save because the local storage quota has been exceeded: you may need to clear out your web browser's local storage for this app, or  run in private mode before you can save again. Specific error here: ";
+
+export interface SetStorageErrorObject {
+  isQuotaError: boolean;
+  message: String;
+  storedStrings: StoredStringDescriptor[];
+}
+
+export interface StoredStringDescriptor {
+  key: string;
+  storageUsedInMB: number;
+}
+
+function getEmptySetStorageErrorObject(): SetStorageErrorObject {
+  return {
+    isQuotaError: false,
+    message: "",
+    storedStrings: [],
+  };
+}
+
+export function setLocalStorageItem(
+  name: string,
+  content: string,
+): [boolean, SetStorageErrorObject] {
+  try {
+    window.localStorage.setItem(name, content);
+    return [true, getEmptySetStorageErrorObject()];
+  } catch (error: any) {
+    // Unfortunately it looks like we can't use typeof or instanceof to determine this one:
+    // all we get is that it's an object so if someone changess the wording we may have to add new checks
+    const isStorageQuotaError =
+      error.stack && error.stack.includes("exceeded the quota.");
+    if (isStorageQuotaError) {
+      return [
+        false,
+        {
+          isQuotaError: true,
+          message: storageQuotaErrorMessage + error,
+          storedStrings: getStoredStrings(),
+        },
+      ];
+    }
+
+    return [false, getEmptySetStorageErrorObject()];
+  }
+}
+
+function getStoredStrings(): StoredStringDescriptor[] {
+  const itemsObject = { ...localStorage };
```

Comment:
This does seem to work, but it doesn't seem documented at https://developer.mozilla.org/en-US/docs/Web/API/Storage. Alternatively there's `length` and `key`, like https://github.com/dabreegster/ltn/blob/47187e79fb5bcddc823a6224a1f80d39d624f9b2/web/src/title/TitleMode.svelte#L30